pkg/storage/cloudimpl/s3_storage.go

```diff
@@ -174,11 +183,32 @@ func (s *s3Storage) WriteFile(ctx context.Context, basename string, content io.R
 	err := contextutil.RunWithTimeout(ctx, "put s3 object",
 		timeoutSetting.Get(&s.settings.SV),
 		func(ctx context.Context) error {
-			_, err := s.s3.PutObjectWithContext(ctx, &s3.PutObjectInput{
+			putObjectInput := s3.PutObjectInput{
 				Bucket: s.bucket,
 				Key:    aws.String(path.Join(s.prefix, basename)),
 				Body:   content,
-			})
+			}
+
+			// If a server side encryption mode is provided in the URI, we must set
+			// the header values to enable SSE before writing the file to the s3
+			// bucket.
+			if s.conf.ServerEncMode != "" {
+				switch s.conf.ServerEncMode {
+				case string(aes256Enc):
+					putObjectInput.SetServerSideEncryption(s.conf.ServerEncMode)
+				case string(kmsEnc):
+					if s.conf.ServerKMSID == "" {
```

Comment:
does this sort of validation feel like it might belong in `MakeS3Storage` instead? seems like `Put` is kinda late to decide the config is invalid?

Reply:
done, moved this, and checking that the mode is one we support, to MakeS3Storage.